File: no-interaction-win/NoInteraction/Core/ApproverEngine.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.ComponentModel;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Threading;
using System.Windows;
using NoInteraction.Models;

namespace NoInteraction.Core
{
    /// <summary>
    /// Windows port of the Mac build's ApproverEngine: owns settings, rules, the activity
    /// log, and the periodic scan loop. Intentionally does not include the Mac app's
    /// Prompt Queue / Loop Mode auto-paste feature.
    /// </summary>
    public sealed class ApproverEngine : INotifyPropertyChanged, IDisposable
    {
        public static readonly ApproverEngine Shared = new();

        public event PropertyChangedEventHandler? PropertyChanged;
        private void Raise([CallerMemberName] string? name = null) =>
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(name));

        private readonly SettingsStore _settings;

        private bool _isEnabled;
        public bool IsEnabled
        {
            get => _isEnabled;
            set { if (_isEnabled == value) return; _isEnabled = value; _settings.IsEnabled = value; _settings.Save(); Raise(); }
        }

        // Windows has no OS-level "grant accessibility" gate for same-privilege automation,
        // so this always reports true; kept for structural parity with the Mac dashboard.
        public bool IsAccessibilityGranted => true;

        private int _totalApprovalsCount;
        public int TotalApprovalsCount
        {
            get => _totalApprovalsCount;
            private set { _totalApprovalsCount = value; _settings.TotalApprovalsCount = value; _settings.Save(); Raise(); }
        }

        private bool _soundEnabled;
        public bool SoundEnabled
        {
            get => _soundEnabled;
            set { if (_soundEnabled == value) return; _soundEnabled = value; _settings.SoundEnabled = value; _settings.Save(); Raise(); }
        }

        public ObservableCollection<LogEntry> Logs { get; } = new();
        public ObservableCollection<ApprovalRule> ButtonRules { get; } = new();
        public ObservableCollection<ApprovalRule> CheckboxRules { get; } = new();

        public static readonly string DefaultPrompt = @"Perform a complete, exhaustive, and uncompromising security, architecture, performance, and UI/UX audit of this entire codebase. Analyze every single line of code with extreme depth and rigor.

Your objective is to optimize this application to the absolute highest tier of software quality in existence. Follow these strict directives:
1. BUG DETECTION & RESOLUTION: Scan for any logical bugs, concurrency race conditions, memory leaks, performance bottlenecks, edge-case crashes, and API misuses. Resolve them immediately with clean, production-ready, and robust code.
2. CODE OPTIMIZATION & REFACTORING: Optimize compile times, memory footprints, and CPU utilization. Eliminate redundant loops and heavy UI renderings. Ensure optimal concurrency paradigms.
3. UI/UX REFINEMENT: Review all layouts, fonts, spacing, color contrasts, transitions, and hover animations. Upgrade the visual design system to feel premium, modern, and state-of-the-art.
4. EDGE CASES & ROBUSTNESS: Ensure perfect error handling, validation, and defensive coding against unexpected window hierarchies or missing permissions.
5. DEEP SEARCH: Use the internet, latest documentation, SDK guidelines, and the full extent of your cognitive capacity. Do not stop until this codebase is completely flawless.";

        public ObservableCollection<string> PromptQueue { get; } = new();

        private int _currentPromptIndex;
        public int CurrentPromptIndex
        {
            get => _currentPromptIndex;
            set { if (_currentPromptIndex == value) return; _currentPromptIndex = value; _settings.CurrentPromptIndex = value; _settings.Save(); Raise(); }
        }

        private bool _isPromptQueueActive = true;
        public bool IsPromptQueueActive
        {
            get => _isPromptQueueActive;
            set { if (_isPromptQueueActive == value) return; _isPromptQueueActive = value; _settings.IsPromptQueueActive = value; _settings.Save(); Raise(); }
        }

        private bool _loopModeEnabled;
        public bool LoopModeEnabled
        {
            get => _loopModeEnabled;
            set { if (_loopModeEnabled == value) return; _loopModeEnabled = value; _settings.LoopModeEnabled = value; _settings.Save(); Raise(); }
        }

        private int _loopModeLimit = 10;
        public int LoopModeLimit
        {
            get => _loopModeLimit;
            set { if (_loopModeLimit == value) return; _loopModeLimit = value; _settings.LoopModeLimit = value; _settings.Save(); Raise(); }
        }

        private int _loopModeCounter;
        public int LoopModeCounter
        {
            get => _loopModeCounter;
            set { if (_loopModeCounter == value) return; _loopModeCounter = value; _settings.LoopModeCounter = value; _settings.Save(); Raise(); }
        }

        private bool _terminalMonitoringEnabled = true;
        public bool TerminalMonitoringEnabled
        {
            get => _terminalMonitoringEnabled;
            set { if (_terminalMonitoringEnabled == value) return; _terminalMonitoringEnabled = value; Raise(); }
        }

        public ObservableCollection<TerminalSession> TerminalSessions { get; } = new();

        private static readonly string[] DefaultButtons =
        {
            "Submit", "Allow", "Always Allow", "Allow All", "Yes, allow", "Yes, and always", "Approve",
            "Yes", "Confirm", "Proceed", "Accept", "Continue", "OK", "Trust", "Got it", "Install", "Open",
            "Run Command", "Run", "Execute", "Always Allow Command", "Always Run", "Run Tool", "Allow Tool"
        };
        private static readonly string[] DefaultCheckboxes =
        {
            "Remember", "Always", "Trust", "Don't ask", "Don't show", "Remember my choice", "Do not ask again"
        };

        private readonly System.Threading.Timer _timer;
        private DateTime _lastActionTime = DateTime.MinValue;
        private readonly TimeSpan _cooldown = TimeSpan.FromSeconds(1.2);
        private volatile bool _ocrScanInFlight;
        private readonly object _scanLock = new();

        private ApproverEngine()
        {
            _settings = SettingsStore.Load();

            _isEnabled = _settings.IsEnabled;
            _soundEnabled = _settings.SoundEnabled;
            _totalApprovalsCount = _settings.TotalApprovalsCount;

            var buttons = _settings.ButtonRules.Count > 0
                ? _settings.ButtonRules
                : DefaultButtons.Select(k => new ApprovalRule(k, TargetType.Button)).ToList();
            foreach (var r in buttons) ButtonRules.Add(r);

            var checkboxes = _settings.CheckboxRules.Count > 0
                ? _settings.CheckboxRules
                : DefaultCheckboxes.Select(k => new ApprovalRule(k, TargetType.Checkbox)).ToList();
            foreach (var r in checkboxes) CheckboxRules.Add(r);

            if (_settings.PromptQueue.Count > 0)
            {
                foreach (var pq in _settings.PromptQueue) PromptQueue.Add(pq);
            }
            else
            {
                PromptQueue.Add(DefaultPrompt);
            }
            _currentPromptIndex = _settings.CurrentPromptIndex;
            _isPromptQueueActive = _settings.IsPromptQueueActive;
            _loopModeEnabled = _settings.LoopModeEnabled;
            _loopModeLimit = _settings.LoopModeLimit;
            _loopModeCounter = _settings.LoopModeCounter;

            ButtonRules.CollectionChanged += (_, _) => SaveRules();
            CheckboxRules.CollectionChanged += (_, _) => SaveRules();
            PromptQueue.CollectionChanged += (_, _) => SavePromptQueue();

            _timer = new System.Threading.Timer(_ => ScheduleScan(), null, TimeSpan.FromSeconds(1), TimeSpan.FromSeconds(3));
        }

        private void SaveRules()
        {
            _settings.ButtonRules = ButtonRules.ToList();
            _settings.CheckboxRules = CheckboxRules.ToList();
            _settings.Save();
        }

        public void SavePromptQueue()
        {
            _settings.PromptQueue = PromptQueue.ToList();
            _settings.Save();
        }

        public void ResetPromptQueueToDefault()
        {
            PromptQueue.Clear();
            PromptQueue.Add(DefaultPrompt);
            CurrentPromptIndex = 0;
            IsPromptQueueActive = true;
        }

        // MARK: Scan loop

        private void ScheduleScan()
        {
            if (!IsEnabled) return;
            if (DateTime.Now - _lastActionTime < _cooldown) return;

            var targetApps = AppObserver.Shared.FindTargetApplications();
            if (targetApps.Count == 0)
            {
                // Slow down scan frequency when no target applications are running
                _timer.Change(TimeSpan.FromSeconds(3.5), TimeSpan.FromSeconds(3.5));
                return;
            }

            // Speed up scan frequency (1s) when target applications are active
            _timer.Change(TimeSpan.FromSeconds(1.0), TimeSpan.FromSeconds(1.0));

            List<string> buttons = new();
            List<string> checkboxes = new();
            Application.Current?.Dispatcher.Invoke(() =>
            {
                buttons = ButtonRules.Where(r => r.IsEnabled).Select(r => r.Keyword).ToList();
                checkboxes = CheckboxRules.Where(r => r.IsEnabled).Select(r => r.Keyword).ToList();
            });

            if (buttons.Count == 0) return;

            foreach (var app in targetApps)
            {
                string appName;
                try { appName = string.IsNullOrEmpty(app.MainWindowTitle) ? app.ProcessName : app.MainWindowTitle; }
                catch { appName = "Target App"; }

                UiaInspector.InspectionResult? result;
                try { result = UiaInspector.Shared.InspectAndAutoApprove(app, buttons, checkboxes); }
                catch { result = null; }

                if (result != null)
                {
                    _lastActionTime = DateTime.Now;
                    if (result.Action == "Fallback Click Needed" && result.Position.HasValue)
                    {
                        ClickAutomation.Shared.PerformClick(result.Position.Value, () =>
                            Record(appName, result.ElementText, "UIA + Click"));
                    }
                    else
                    {
                        Record(appName, result.ElementText, result.Action);
                    }
                    return; // matches Mac behavior: stop after the first successful action this tick
                }
            }

            // Pass 2: OCR fallback if nothing was found via UI Automation
            if (_ocrScanInFlight) return;
            foreach (var app in targetApps)
            {
                var bounds = AppObserver.Shared.GetWindowBounds(app);
                if (bounds == null) continue;

                _ocrScanInFlight = true;
                var capturedApp = app;
                string capturedAppName = "Target App";
                try { capturedAppName = string.IsNullOrEmpty(app.MainWindowTitle) ? app.ProcessName : app.MainWindowTitle; } catch { }

                _ = RunOcrPassAsync(bounds.Value, buttons, capturedApp, capturedAppName);
                break;
            }
        }

        public void Dispose()
        {
            _timer.Dispose();
        }

        private async System.Threading.Tasks.Task RunOcrPassAsync(Rect bounds, List<string> buttons, System.Diagnostics.Process app, string appName)
        {
            try
            {
                var (point, text) = await OcrScanner.Shared.ScanRegionForKeywordsAsync(bounds, buttons);
                if (point == null || text == null) return;
                if (DateTime.Now - _lastActionTime < _cooldown) return;

                _lastActionTime = DateTime.Now;
                ClickAutomation.Shared.PerformClick(point.Value, () => Record(appName, text, "OCR"));
            }
            finally
            {
                _ocrScanInFlight = false;
            }
        }

        // MARK: Logging & audio feedback

        private void Record(string appName, string text, string method)
        {
            Application.Current?.Dispatcher.Invoke(() =>
            {
                TotalApprovalsCount += 1;
                var entry = new LogEntry
                {
                    AppName = appName,
                    ActionTaken = "Auto-Approved",
                    TargetText = text,
                    DetectionMethod = method
                };
                Logs.Insert(0, entry);
                while (Logs.Count > 200) Logs.RemoveAt(Logs.Count - 1);

                if (SoundEnabled)
                {
                    try { System.Media.SystemSounds.Asterisk.Play(); } catch { }
                }
            });
        }

        // MARK: Rule management

        public void AddRule(string keyword, TargetType targetType)
        {
            var kw = keyword.Trim();
            if (string.IsNullOrEmpty(kw)) return;
            var collection = targetType == TargetType.Button ? ButtonRules : CheckboxRules;
            if (collection.Any(r => string.Equals(r.Keyword, kw, StringComparison.OrdinalIgnoreCase))) return;
            collection.Add(new ApprovalRule(kw, targetType));
        }

        public void RemoveRule(Guid id, TargetType targetType)
        {
            var collection = targetType == TargetType.Button ? ButtonRules : CheckboxRules;
            var rule = collection.FirstOrDefault(r => r.Id == id);
            if (rule != null) collection.Remove(rule);
        }

        public void ToggleRule(Guid id, TargetType targetType)
        {
            var collection = targetType == TargetType.Button ? ButtonRules : CheckboxRules;
            var rule = collection.FirstOrDefault(r => r.Id == id);
            if (rule == null) return;
            rule.IsEnabled = !rule.IsEnabled;
            SaveRules();
            // Force UI refresh since mutating a property on an item doesn't raise CollectionChanged.
            var idx = collection.IndexOf(rule);
            collection.RemoveAt(idx);
            collection.Insert(idx, rule);
        }

        public void ResetRulesToDefault()
        {
            ButtonRules.Clear();
            foreach (var k in DefaultButtons) ButtonRules.Add(new ApprovalRule(k, TargetType.Button));
            CheckboxRules.Clear();
            foreach (var k in DefaultCheckboxes) CheckboxRules.Add(new ApprovalRule(k, TargetType.Checkbox));
        }
    }
}

File: no-interaction-mac/NoInteraction/Core/ApproverEngine.swift
import Cocoa
import Combine

@MainActor
public final class ApproverEngine: ObservableObject {
    public static let shared = ApproverEngine()

    // MARK: - Published State

    @Published public var isEnabled: Bool = true {
        didSet { UserDefaults.standard.set(isEnabled, forKey: "isEnabled") }
    }
    @Published public var isAccessibilityGranted: Bool = false
    @Published public var totalApprovalsCount: Int = 0 {
        didSet { UserDefaults.standard.set(totalApprovalsCount, forKey: "totalApprovalsCount") }
    }
    @Published public var soundEnabled: Bool = true {
        didSet { UserDefaults.standard.set(soundEnabled, forKey: "soundEnabled") }
    }
    @Published public var logs: [LogEntry] = []

    @Published public var buttonRules: [ApprovalRule] = [] {
        didSet { saveRules() }
    }

    @Published public var checkboxRules: [ApprovalRule] = [] {
        didSet { saveRules() }
    }

    @Published public var promptQueue: [String] = [] {
        didSet { savePromptQueue() }
    }
    @Published public var currentPromptIndex: Int = 0 {
        didSet { UserDefaults.standard.set(currentPromptIndex, forKey: "currentPromptIndex") }
    }
    @Published public var isPromptQueueActive: Bool = false {
        didSet { UserDefaults.standard.set(isPromptQueueActive, forKey: "isPromptQueueActive") }
    }
    @Published public var loopModeEnabled: Bool = false {
        didSet { UserDefaults.standard.set(loopModeEnabled, forKey: "loopModeEnabled") }
    }
    @Published public var loopModeLimit: Int = 10 {
        didSet { UserDefaults.standard.set(loopModeLimit, forKey: "loopModeLimit") }
    }
    @Published public var loopModeCounter: Int = 0 {
        didSet { UserDefaults.standard.set(loopModeCounter, forKey: "loopModeCounter") }
    }

    // MARK: - Private State

    private var timer: Timer?
    private var lastActionTime: Date = .distantPast
    private let cooldown: TimeInterval = 1.2
    private var visionScanInFlight = false

    // Default Fallback Rules
    private static let defaultButtons = [
        "Submit", "Yes, allow this time", "Yes, allow", "Allow running this command",
        "Allow", "Yes, and always", "Approve", "Yes", "Confirm", "Proceed",
        "Accept", "Continue", "OK"
    ]
    private static let defaultCheckboxes = [
        "Remember", "Always", "Trust", "Don't ask", "Don't show"
    ]

    public static let defaultPrompt = """
Perform a complete, exhaustive, and uncompromising security, architecture, performance, and UI/UX audit of this entire codebase. Analyze every single line of code with extreme depth and rigor.

Your objective is to optimize this application to the absolute highest tier of software quality in existence. Follow these strict directives:
1. BUG DETECTION & RESOLUTION: Scan for any logical bugs, concurrency race conditions, memory leaks, performance bottlenecks, edge-case crashes, and API misuses. Resolve them immediately with clean, production-ready, and robust code.
2. CODE OPTIMIZATION & REFACTORING: Optimize compile times, memory footprints, and CPU utilization. Eliminate redundant loops, redundant accessibility calls, and heavy UI renderings. Ensure optimal Swift concurrency paradigms.
3. UI/UX REFINEMENT: Review all layouts, fonts, spacing, color contrasts, transitions, and hover animations. Upgrade the visual design system to feel premium, modern, and state-of-the-art.
4. EDGE CASES & ROBUSTNESS: Ensure perfect error handling, validation, and defensive coding against unexpected window hierarchies, missing permissions, or browser states.
5. DEEP SEARCH: Use the internet, latest documentation, Apple SDK guidelines, and the full extent of your cognitive capacity. Do not stop until this codebase is completely flawless.
"""

    private init() {
        loadSettingsAndRules()
        refreshAccessibilityStatus()
        startTimer()
        setupWorkspaceNotifications()
    }

    // MARK: - Persistence

    private func loadSettingsAndRules() {
        let defaults = UserDefaults.standard
        if defaults.object(forKey: "isEnabled") != nil {
            self.isEnabled = defaults.bool(forKey: "isEnabled")
        }
        self.totalApprovalsCount = defaults.integer(forKey: "totalApprovalsCount")
        if defaults.object(forKey: "soundEnabled") != nil {
            self.soundEnabled = defaults.bool(forKey: "soundEnabled")
        } else {
            self.soundEnabled = true
        }

        // Load Rules
        if let data = defaults.data(forKey: "buttonRules"),
           let decoded = try? JSONDecoder().decode([ApprovalRule].self, from: data) {
            self.buttonRules = decoded
        } else {
            self.buttonRules = ApproverEngine.defaultButtons.map {
                ApprovalRule(keyword: $0, targetType: .button)
            }
        }
        // Sanitize rules: remove Run and Execute to prevent conflict with IDE "Run Code" buttons
        self.buttonRules.removeAll { $0.keyword.caseInsensitiveCompare("Run") == .orderedSame || $0.keyword.caseInsensitiveCompare("Execute") == .orderedSame }

        if let data = defaults.data(forKey: "checkboxRules"),
           let decoded = try? JSONDecoder().decode([ApprovalRule].self, from: data) {
            self.checkboxRules = decoded
        } else {
            self.checkboxRules = ApproverEngine.defaultCheckboxes.map {
                ApprovalRule(keyword: $0, targetType: .checkbox)
            }
        }

        if let pq = defaults.stringArray(forKey: "promptQueue"), !pq.isEmpty {
            self.promptQueue = pq
        } else {
            self.promptQueue = [ApproverEngine.defaultPrompt]
            self.isPromptQueueActive = true
        }
        self.currentPromptIndex = defaults.integer(forKey: "currentPromptIndex")
        self.isPromptQueueActive = defaults.object(forKey: "isPromptQueueActive") != nil ? defaults.bool(forKey: "isPromptQueueActive") : true
        self.loopModeEnabled = defaults.bool(forKey: "loopModeEnabled")
        self.loopModeLimit = defaults.object(forKey: "loopModeLimit") != nil ? defaults.integer(forKey: "loopModeLimit") : 10
        self.loopModeCounter = defaults.integer(forKey: "loopModeCounter")
    }

    private func saveRules() {
        if let bData = try? JSONEncoder().encode(buttonRules) {
            UserDefaults.standard.set(bData, forKey: "buttonRules")
        }
        if let cData = try? JSONEncoder().encode(checkboxRules) {
            UserDefaults.standard.set(cData, forKey: "checkboxRules")
        }
    }

    private func savePromptQueue() {
        UserDefaults.standard.set(promptQueue, forKey: "promptQueue")
    }

    // MARK: - Accessibility

    public func refreshAccessibilityStatus() {
        let options: NSDictionary = [kAXTrustedCheckOptionPrompt.takeUnretainedValue() as String: false]
        let granted = AXIsProcessTrustedWithOptions(options)
        if granted != self.isAccessibilityGranted {
            self.isAccessibilityGranted = granted
        }
    }

    public func promptAccessibilityPermission() {
        let options: NSDictionary = [kAXTrustedCheckOptionPrompt.takeUnretainedValue() as String: true]
        _ = AXIsProcessTrustedWithOptions(options)
        DispatchQueue.main.asyncAfter(deadline: .now() + 1.5) { [weak self] in
            Task { @MainActor in
                self?.refreshAccessibilityStatus()
            }
        }
    }

    // MARK: - Scan Loop

    public func startTimer() {
        timer?.invalidate()
        // Run checks every 1.5 seconds for improved responsiveness, scaling up when app is active
        timer = Timer.scheduledTimer(withTimeInterval: 1.5, repeats: true) { [weak self] _ in
            Task { @MainActor in
                self?.scheduleScan()
            }
        }
        RunLoop.main.add(timer!, forMode: .common)
    }

    private func setupWorkspaceNotifications() {
        NSWorkspace.shared.notificationCenter.addObserver(
            self,
            selector: #selector(handleWorkspaceNotification),
            name: NSWorkspace.didActivateApplicationNotification,
            object: nil
        )
    }

    @objc private func handleWorkspaceNotification(_ notification: Notification) {
        // Immediate scan on app switch
        scheduleScan()
    }

    private func scheduleScan() {
        refreshAccessibilityStatus()
        guard isEnabled, isAccessibilityGranted else { return }
        guard Date().timeIntervalSince(lastActionTime) >= cooldown else { return }

        let buttons    = buttonRules.filter(\.isEnabled).map(\.keyword)
        let checkboxes = checkboxRules.filter(\.isEnabled).map(\.keyword)
        guard !buttons.isEmpty else { return }

        let targetApps = AppObserver.shared.findTargetApplications()
        guard !targetApps.isEmpty else { return }

        print("🔄 scheduleScan running. Target apps found: \(targetApps.map { $0.localizedName ?? "" })")

        // Perform scan sequentially for each target application
        for app in targetApps {
            let pid = app.processIdentifier
            let appName = app.localizedName ?? "Anti-Gravity"

            Task {
                // Pass 1: AXPress / AXClick (runs tree traversal in detached background task)
                let result = await Task.detached(priority: .userInitiated) {
                    return AXInspector.shared.inspectAndAutoApprove(
                        pid: pid,
                        buttonKeywords: buttons,
                        checkboxKeywords: checkboxes
                    )
                }.value

                if let result = result {
                    self.lastActionTime = Date()
                    if result.action == "Fallback CGEvent Click Needed", let pt = result.position {
                        ClickAutomation.shared.performClick(at: pt, targetPid: pid) {
                            Task { @MainActor in
                                self.record(appName: appName, text: result.elementText, method: "AX + CGClick")
                            }
                        }
                    } else {
                        self.record(appName: appName, text: result.elementText, method: result.action)
                    }
                    return
                }

                // Pass 2: Vision OCR (Fallback if AX is blocked or missed)
                if !self.visionScanInFlight, let bounds = AppObserver.shared.getWindowBounds(for: app) {
                    self.visionScanInFlight = true
                    let capturedPid = pid

                    let ocrResult = await withCheckedContinuation { (continuation: CheckedContinuation<(CGPoint?, String?), Never>) in
                        VisionOCRScanner.shared.scanRegionForKeywords(
                            windowBounds: bounds,
                            buttonKeywords: buttons
                        ) { point, text in
                            continuation.resume(returning: (point, text))
                        }
                    }
                    
                    self.visionScanInFlight = false
                    if let pt = ocrResult.0, let text = ocrResult.1 {
                        if Date().timeIntervalSince(self.lastActionTime) >= self.cooldown {
                            self.lastActionTime = Date()
                            ClickAutomation.shared.performClick(at: pt, targetPid: capturedPid) {
                                Task { @MainActor in
                                    self.record(appName: appName, text: text, method: "Vision OCR")
                                }
                            }
                        }
                        return
                    }
                }

                // Pass 3: Prompt Queue Check or Loop Mode Check (ONLY if no approval buttons were found)
                let shouldCheckPromptState = (self.isPromptQueueActive && self.currentPromptIndex < self.promptQueue.count) ||
                                             (self.loopModeEnabled && (self.loopModeLimit == 0 || self.loopModeCounter < self.loopModeLimit))

                if shouldCheckPromptState {
                    let chatState = await Task.detached(priority: .userInitiated) { () -> AXInspector.ChatStateResult in
                        let axApp = AXUIElementCreateApplication(pid)
                        var windowsRef: CFTypeRef?
                        var isBusy = false
                        var inputArea: AXUIElement? = nil
                        
                        if AXUIElementCopyAttributeValue(axApp, kAXWindowsAttribute as CFString, &windowsRef) == .success,
                           let windows = windowsRef as? [AXUIElement] {
                            let filteredWindows = AppObserver.shared.isBrowser(app)
                                ? windows.filter { AppObserver.shared.isAntigravityWindow($0) }
                                : windows
                            
                            for win in filteredWindows {
                                let state = AXInspector.shared.findChatInputAndStatus(in: win, depth: 0)
                                if state.isBusy {
                                    isBusy = true
                                }
                                if let input = state.inputArea {
                                    inputArea = input
                                }
                            }
                        }
                        return AXInspector.ChatStateResult(inputArea: inputArea, isBusy: isBusy)
                    }.value

                    if !chatState.isBusy, let input = chatState.inputArea {
                        let prompt = self.loopModeEnabled ? ApproverEngine.defaultPrompt : self.promptQueue[self.currentPromptIndex]
                        await self.pastePrompt(prompt: prompt, toInput: input, forApp: app)
                        return
                    }
                }
            }
        }
    }

    private func pastePrompt(prompt: String, toInput input: AXUIElement, forApp app: NSRunningApplication) async {
        if self.loopModeEnabled {
            self.loopModeCounter += 1
            print("🤖 Loop Mode: Pasting prompt \(self.loopModeCounter)/\(self.loopModeLimit == 0 ? "∞" : String(self.loopModeLimit))")
        } else {
            self.currentPromptIndex += 1
            print("🤖 Prompt Queue: Pasting prompt \(self.currentPromptIndex)/\(self.promptQueue.count)")
        }
        self.lastActionTime = Date()

        app.activate(options: .activateIgnoringOtherApps)
        try? await Task.sleep(nanoseconds: 200_000_000) // 0.2s

        if let center = AXInspector.shared.centerOf(input) {
            print("🎯 Clicking center of chat input area at: \(center) for app: \(app.localizedName ?? "")")
            await withCheckedContinuation { (continuation: CheckedContinuation<Void, Never>) in
                ClickAutomation.shared.performClick(at: center, targetPid: app.processIdentifier) {
                    continuation.resume()
                }
            }

            try? await Task.sleep(nanoseconds: 250_000_000) // 0.25s

            NSPasteboard.general.clearContents()
            NSPasteboard.general.setString(prompt, forType: .string)
            try? await Task.sleep(nanoseconds: 150_000_000) // 0.15s

            print("📋 Pasting prompt...")
            AXInspector.shared.pressPasteKeystroke()
            try? await Task.sleep(nanoseconds: 200_000_000) // 0.2s

            print("⌨️ Pressing Return key...")
            AXInspector.shared.pressReturnKey()
        }
    }

    // MARK: - Logging & Audio Feedback

    /// Entry point for other engines (e.g. `TerminalApproverEngine`) to share the same
    /// Activity Log and sound feedback instead of keeping a separate log per source.
    public func recordExternal(appName: String, text: String, method: String) {
        record(appName: appName, text: text, method: method)
    }

    private func record(appName: String, text: String, method: String) {
        self.totalApprovalsCount += 1
        let entry = LogEntry(
            appName: appName,
            actionTaken: "Auto-Approved",
            targetText: text,
            detectionMethod: method
        )
        self.logs.insert(entry, at: 0)
        if self.logs.count > 200 { self.logs = Array(self.logs.prefix(200)) }
        MenuBarManager.shared.rebuildMenu()

        if self.soundEnabled {
            NSSound(named: "Tink")?.play()
        }
    }

    // MARK: - Rule Management

    public func addRule(keyword: String, targetType: ApprovalRule.TargetType) {
        let kw = keyword.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !kw.isEmpty else { return }
        let rule = ApprovalRule(keyword: kw, targetType: targetType)
        if targetType == .button {
            guard !buttonRules.contains(where: { $0.keyword.caseInsensitiveCompare(kw) == .orderedSame }) else { return }
            buttonRules.append(rule)
        } else {
            guard !checkboxRules.contains(where: { $0.keyword.caseInsensitiveCompare(kw) == .orderedSame }) else { return }
            checkboxRules.append(rule)
        }
    }

    public func removeRule(id: UUID, targetType: ApprovalRule.TargetType) {
        if targetType == .button {
            buttonRules.removeAll { $0.id == id }
        } else {
            checkboxRules.removeAll { $0.id == id }
        }
    }

    public func toggleRule(id: UUID, targetType: ApprovalRule.TargetType) {
        if targetType == .button {
            if let idx = buttonRules.firstIndex(where: { $0.id == id }) {
                buttonRules[idx].isEnabled.toggle()
            }
        } else {
            if let idx = checkboxRules.firstIndex(where: { $0.id == id }) {
                checkboxRules[idx].isEnabled.toggle()
            }
        }
    }

    public func resetRulesToDefault() {
        buttonRules = ApproverEngine.defaultButtons.map { ApprovalRule(keyword: $0, targetType: .button) }
        checkboxRules = ApproverEngine.defaultCheckboxes.map { ApprovalRule(keyword: $0, targetType: .checkbox) }
    }

    public func resetPromptQueueToDefault() {
        promptQueue = [ApproverEngine.defaultPrompt]
        currentPromptIndex = 0
        isPromptQueueActive = true
    }
}
